### ISSUE TITLE
problem with unearth package on Windows

### ISSUE THREAD

**Issue by AdamJel** [pdm-project/pdm#3115]
- [x] I have searched the issue tracker and believe that this is not a duplicate.

I am trying to use the `pdm sync` command to "install" locally available packages in form of wheel files to a current project. The command fails with error, first because of a problem with the `unearth` package, and then because it tries to contact internet services, which are not available since the machine does not have internet connection.

The whole procedure is to prepare everything for off-line installation on a dev machine, copy prepared files over to a remote machine and then "install" dependencies with the `sync` command.

## Steps to reproduce

- download pdm wheel with `pdm run python -m pip download --only-binary :all: --platform win_amd64 --require-hashes -d packages/pdm`, which fills the directory with wheel files of pdm and its dependencies including the `unearth` package.
- on a target Windows machine, install pdm via pip: `python -m pip install --no-index --find-links=. pdm`
- include the following snippet in `pyproject.toml` for enabling local install of packages (with actual path):
  ```
  [tool.pdm.source]
  name = "local"
  url = "file:///${PROJECT_ROOT}/packages"
  type = "find_links"
  ```
- lock dependencies of the project `pdm lock --prod --python ==3.12.* --platform windows_amd64 --implementation cpython --lockfile pdm.remote.lock`
- export these dependencies to requirements file `pdm export --prod --lockfile pdm.remote.lock -f requirements -o requirements.remote.txt`
- download these dependencies with pip `pdm run python -m pip download -r requirements.remote.txt --only-binary :all: --platform win_amd64 --require-hashes -d packages`
- copy them all over to remote Windows machine, alongside with the pdm.lock file
- on remote Win machine, "install" them with `pdm sync --clean --reinstall --prod`

## Actual behavior

The sync command `pdm sync --clean --reinstall --prod -v` produces the following error:

```stdout
pdm : WARNING: Lockfile hash doesn't match pyproject.toml, packages may be outdated

STATUS: Resolving packages from lockfile...
unearth.preparer: The file . is a local directory, use it directly
Synchronizing working set with resolved packages: 10 to add, 0 to update, 0 to remove

unearth.utils: Couldn't parse netrc because of FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\ELVAC\\.netrc'
unearth.utils: Couldn't parse netrc because of FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\ELVAC\\.netrc'
unearth.utils: Couldn't parse netrc because of FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\ELVAC\\.netrc'
unearth.utils: Couldn't parse netrc because of FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\ELVAC\\.netrc'
unearth.utils: Couldn't parse netrc because of FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\ELVAC\\.netrc'
unearth.utils: Couldn't parse netrc because of FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\ELVAC\\.netrc'
unearth.utils: Couldn't parse netrc because of FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\ELVAC\\.netrc'
unearth.utils: Couldn't parse netrc because of FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\ELVAC\\.netrc'
  x Install greenlet 3.0.3 failed
  x Install pyodbc 5.1.0 failed
  x Install six 1.16.0 failed
  x Install python-dateutil 2.9.0.post0 failed
  x Install sqlalchemy 2.0.32 failed
  x Install numpy 2.0.1 failed
  x Install pandas 2.2.2 failed
  x Install pytz 2024.1 failed
```

The stdout continues with the same traceback for each dependency:

```stdout
pdm.termui: Error occurs adding greenlet: 
Traceback (most recent call last):
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpx\_transports\default.py", line 69, in 
map_httpcore_exceptions
    yield
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpx\_transports\default.py", line 233, in 
handle_request
    resp = self._pool.handle_request(req)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpcore\_sync\connection_pool.py", line 216, in 
handle_request
    raise exc from None
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpcore\_sync\connection_pool.py", line 196, in 
handle_request
    response = connection.handle_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpcore\_sync\connection.py", line 99, in 
handle_request
    raise exc
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpcore\_sync\connection.py", line 76, in 
handle_request
    stream = self._connect(request)
             ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpcore\_sync\connection.py", line 122, in _connect
    stream = self._network_backend.connect_tcp(**kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpcore\_backends\sync.py", line 205, in connect_tcp
    with map_exceptions(exc_map):
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\contextlib.py", line 158, in __exit__
    self.gen.throw(value)
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpcore\_exceptions.py", line 14, in map_exceptions
    raise to_exc(exc) from exc
httpcore.ConnectError: [Errno 11001] getaddrinfo failed

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\concurrent\futures\thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\pdm\installers\synchronizers.py", line 228, in 
install_candidate
    self.manager.install(can)
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\pdm\installers\manager.py", line 33, in install
    prepared.build(),
    ^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\pdm\models\candidates.py", line 392, in build
    self._obtain(allow_all=False)
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\pdm\models\candidates.py", line 429, in _obtain
    self.link = _find_best_match_link(
                ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\pdm\models\candidates.py", line 91, in 
_find_best_match_link
    best = finder.find_best_match(req.as_line(), hashes=hashes).best
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\unearth\finder.py", line 396, in find_best_match
    best_match = next(iter(applicable_candidates), None)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\unearth\finder.py", line 329, in 
_find_packages_from_requirement
    yield from self._find_packages(requirement.name, allow_yanked)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\unearth\finder.py", line 295, in _find_packages
    return sorted(all_packages, key=self._sort_key, reverse=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\unearth\collector.py", line 187, in 
collect_links_from_location
    yield from _collect_links_from_index(session, location)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\unearth\collector.py", line 210, in 
_collect_links_from_index
    page = fetch_page(session, location, headers)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\unearth\collector.py", line 195, in fetch_page
    resp = _get_html_response(session, location, headers)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\unearth\collector.py", line 234, in 
_get_html_response
    resp = session.get(
           ^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpx\_client.py", line 1054, in get
    return self.request(
           ^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpx\_client.py", line 827, in request
    return self.send(request, auth=auth, follow_redirects=follow_redirects)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpx\_client.py", line 914, in send
    response = self._send_handling_auth(
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpx\_client.py", line 942, in _send_handling_auth
    response = self._send_handling_redirects(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpx\_client.py", line 979, in 
_send_handling_redirects
    response = self._send_single_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpx\_client.py", line 1015, in _send_single_request
    response = transport.handle_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\hishel\_sync\_transports.py", line 213, in 
handle_request
    regular_response = self._transport.handle_request(request)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpx\_transports\default.py", line 232, in 
handle_request
    with map_httpcore_exceptions():
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\contextlib.py", line 158, in __exit__
    self.gen.throw(value)
  File "C:\Users\ELVAC\AppData\Local\Programs\Python\Python312\Lib\site-packages\httpx\_transports\default.py", line 86, in 
map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.ConnectError: [Errno 11001] getaddrinfo failed
```

## Expected behavior

All packages should be installed.

### pdm info

```
PDM version:
  2.18.0
Python Interpreter:
  C:\Users\ELVAC\AppData\Local\thu\mssql\.venv\Scripts\python.exe (3.12)
Project Root:
  C:/Users/ELVAC/AppData/Local/thu/mssql
Local Packages:
```

### pdm info --env

```
{
  "implementation_name": "cpython",
  "implementation_version": "3.12.2",
  "os_name": "nt",
  "platform_machine": "AMD64",
  "platform_release": "10",
  "platform_system": "Windows",
  "platform_version": "10.0.19044",
  "python_full_version": "3.12.2",
  "platform_python_implementation": "CPython",
  "python_version": "3.12",
  "sys_platform": "win32"
}

```

---

## notes

Perhaps there is a problem with where the pdm expects the unearth package to be on Windows?

The bottom line for me is, that I would like and expect the `pdm sync` command to 'just' install the locally available wheels as specified in the lock file without ever contacting network services (internet). To me it seems, that due to the problem with unearth, the pdm tries to resolve the packages against pypi for some reason and fails, because the machine does not have internet connection.

**Comment by frostming** [comment]
On the remote windows server set `pypi.ignore_saved_index` to true.

**Comment by frostming** [comment]
FWIW, you are using `requirements.txt` just for downloading packages, you can use `pdm-download` instead.

**Comment by pawamoy** [comment]
For an offline installation, in a virtualenv, I'd say per-project, with the `--no-isolation` flag.
